File: vol2/ch21/LeetCode1289.py
class Solution:
  def minFallingPathSum(self, grid: List[List[int]]) -> int:
    INF=0x3f3f3f3f
    n=len(grid)
    dp=[[0 for _ in range(0,n)] for _ in range(0,n)]
    for j in range(0,n):  #第0行:边界条件
      dp[0][j]=grid[0][j]
    for i in range(1,n):    #考虑第i行
      for j in range(0,n):  #考虑第i行的第j列
        tmp=INF
        for k in range(0,n):   #考虑第i-1行的第k列
          if k!=j:tmp=min(tmp,dp[i-1][k])
        dp[i][j]=grid[i][j]+tmp
    return min(dp[n-1])   #返回dp[n-1]中的最小值

File: vol2/ch21/LeetCode1289.cpp
class Solution {
  const int INF=0x3f3f3f3f;
public:
  int minFallingPathSum(vector<vector<int>>& grid) {
    int n=grid.size();
    vector<vector<int>> dp=vector<vector<int>>(n,vector<int>(n,0));
    for (int j=0;j<n;j++)             //第0行:边界条件
      dp[0][j]=grid[0][j];
    for (int i=1;i<n;i++) {           //考虑第i行
      for (int j=0;j<n;j++) {         //考虑第i行的第j列
        int tmp=INF;
        for (int k=0;k<n;k++) {       //考虑第i-1行的第k列
          if (k!=j) tmp=min(tmp,dp[i-1][k]);
        }
        dp[i][j]=grid[i][j]+tmp;
      }
    }
    int ans=dp[n-1][0];             //求dp[n-1]中的最小值
    for (int j=1;j<n;j++) {
        ans=min(ans,dp[n-1][j]);
    }
    return ans;
  }
};

